onnx/shape_inference/implementation.cc

```diff
@@ -345,6 +346,56 @@ class ShapeInferenceImplBase {
     }
   }
 
+  // Initialize a DataValueMap for a called function from the DataValueMap of the caller
+  void bindValuesOnCall(
+      const DataValueMap& caller_map,
+      const NodeProto& caller,
+      DataValueMap& callee_map,
+      const FunctionProto& callee) {
+    auto num_inputs = (std::min)(caller.input_size(), callee.input_size());
+    for (int i = 0; i < num_inputs; ++i) {
+      const std::string& actual = caller.input(i);
+      const std::string& formal = callee.input(i);
+      if (!actual.empty()) {
+        auto it = caller_map.find(actual);
+        if (it != caller_map.end()) {
+          callee_map[formal] = it->second;
+        }
+      }
+    }
+  }
+
+  // Update a DataValueMap for a calling function from the DataValueMap of the callee
+  void bindValuesOnReturn(
```

Comment:
The code is very similar to the previous one, maybe that would be worth merging hem.

Reply:
Yes, they are similar. But hard to merge them. One focuses on inputs, the other on outputs. One copies from caller to callee, the other from callee to caller. But, in either case, we need to check caller-side input/output to see if it is empty. We will need to add so many extra-parameters (to merge them) that understanding the logic will be harder. I think it is better in its current form.

Reply:
This is what I had in mind. I probably missed something.

```c++
  void bindValuesOnCallOrReturn(
      const DataValueMap& caller_map,
      const NodeProto& caller,
      bool io,
      DataValueMap& callee_map,
      const FunctionProto& callee) {
    int num_io = io 
      ? (std::min)(caller.input_size(), callee.input_size())
      : (std::min)(caller.output_size(), callee.output_size());
    for (int i = 0; i < num_io; ++i) {
      const std::string& actual = io ? caller.input(i) : caller.output(i);
      if (!actual.empty()) {
        const std::string& formal = io ? callee.input(i) : callee.output(i);
        auto it = caller_map.find(actual);
        if (it != caller_map.end()) {
          callee_map[formal] = it->second;
        }
      }
    }
  }

  // Initialize a DataValueMap for a called function from the DataValueMap of the caller
  void bindValuesOnCall(
      const DataValueMap& caller_map,
      const NodeProto& caller,
      DataValueMap& callee_map,
      const FunctionProto& callee) {
    bindValuesOnCallOrReturn(caller_map, caller, true, callee_map, callee);
  }

  // Update a DataValueMap for a calling function from the DataValueMap of the callee
  void bindValuesOnReturn(
      const DataValueMap& callee_map,
      const FunctionProto& callee,
      DataValueMap& caller_map,
      const NodeProto& caller) {
    bindValuesOnCallOrReturn(callee_map, caller, false, caller_map, callee);
  }
```

Reply:
In the innermost 3 lines, you will also need to change `caller_map` to `io ? caller_map : callee_map`, and similarly for `callee_map`. I prefer to keep the current form, as it is more understandable and is easier to verify is correct when we examine the code.

Reply:
And similarly we need to do the same for `actual` and `formal` in the innermost 3 lines.